### PR TITLE
session: refactor non-transactional DML to make it ready for more DML types

### PR DESCRIPTION
**Pull request by ekexium**
Signed-off-by: ekexium <eke@fastmail.com>

### What problem does this PR solve?

Issue Number: ref #33485

Problem Summary:

To support `update` and `insert into select` later, first refactor the existing code specific to `delete`.

### What is changed and how it works?

1. Refactor
2. A proof-of-concept implementation of non-transactional update to show the refactor works.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -911,7 +911,7 @@ import (
 	LoadDataStmt               "Load data statement"
 	LoadStatsStmt              "Load statistic statement"
 	LockTablesStmt             "Lock tables statement"
-	NonTransactionalDeleteStmt "Non-transactional delete statement"
+	NonTransactionalDMLStmt    "Non-transactional delete statement"
```
Comment (sticnarf) [inline parser/parser.y:914]: The comment should be also changed?

**Comment by sticnarf** [comment]
Emm, is it really just a refactoring? I think it supports `UPDATE` for non-transaction DML, right?

**Comment by ekexium** [comment]
> Emm, is it really just a refactoring? I think it supports `UPDATE` for non-transaction DML, right?

Yes. It's not a full implementation though. I'd prefer merge the part first so that following work on `UPDATE` and `INSERT` could be parallelized, hopefully.

**Review by sticnarf — COMMENTED** [review]
LGTM. Please fix the conflicts.

**Review by you06 — COMMENTED** [review]
rest LGTM

File: session/nontransactional_test.go [hunk 2]
```diff
@@ -65,24 +65,36 @@ func TestNonTransactionalDeleteShardingOnVarchar(t *testing.T) {
...
-				tk.MustQuery(fmt.Sprintf("batch on a limit %d delete from t", batchSize)).Check(testkit.Rows(fmt.Sprintf("%d all succeeded", (tableSize+batchSize-1)/batchSize)))
-				tk.MustQuery("select count(*) from t").Check(testkit.Rows("0"))
+		for _, c := range compositions {
+			for i := 0; i < c.tableSize; i++ {
+				tk.MustExec(fmt.Sprintf("insert into t values ('%d', %d)", i, i*2))
 			}
+			tk.MustQuery(fmt.Sprintf("batch on a limit %d update t set b = b * 2", c.batchSize)).Check(testkit.Rows(fmt.Sprintf("%d all succeeded", (c.tableSize+c.batchSize-1)/c.batchSize)))
+			tk.MustQuery("select coalesce(sum(b), 0) from t").Check(testkit.Rows(fmt.Sprintf("%d", (c.tableSize-1)*c.tableSize*2)))
+			tk.MustQuery(fmt.Sprintf("batch on a limit %d delete from t", c.batchSize)).Check(testkit.Rows(fmt.Sprintf("%d all succeeded", (c.tableSize+c.batchSize-1)/c.batchSize)))
+			tk.MustQuery("select count(*) from t").Check(testkit.Rows("0"))
```
[the remark this answers is not in the document]
Comment (you06) [reply session/nontransactional_test.go:92]: The lines here are too long(not fully displayed on my screen).

[the remark this answers is not in the document]
Comment (ekexium) [reply session/nontransactional_test.go:106]: OK that makes sense. I used to turn on the auto hard wrap. But it always changes existing code that I'd rather not touch in a PR. Sadly we don't have a hard standard for this :(

**Review by you06 — APPROVED** [review]

**Review by sticnarf — APPROVED** [review]

**Pull request merged**